### ISSUE TITLE
[OSGI] : Overriding Classes does not work in latest master 

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#19578]
When we override classes in dotCMS via OSGI, we rely on the Thread activating the bundle's classloader.  I am not sure when this changed, but it seems like the Thread that is loading our OSGI plugins is an isolated felix-only Thread loaded by a felix classloader, not in the dotCMS classloader.  This means that we we inject the classes to override, we are injecting them into the felix and not in the main dotCMS classloader.  This has the effect of preventing the overridden classes from being visible in the overall dotCMS environment.

 

 

  It seems that felix using

**Comment by nollymar** [comment]
**Needs work:** I get this error when I try to upload a plugin: 

```
05-Jan-2021 14:06:42.000 WARNING [FelixStartLevel] org.apache.tika.config.InitializableProblemHandler$3.handleInitializableProblem org.xerial's sqlite-jdbc is not loaded.
Please provide the jar on your classpath to parse sqlite files.
See tika-parsers/pom.xml for the correct version.
14:06:51.659  INFO  AJAX.OSGIAJAX - OSGI Bundle com.dotcms.override-0.2.jar Uploaded
Installing bundle com.dotcms.override / 0.2.0
14:06:53.888  INFO  override.Activator - Injecting: com.dotmarketing.portlets.folders.model.Folder into classloader: ParallelWebappClassLoader
  context: ROOT
  delegate: true
----------> Parent Classloader:
java.net.URLClassLoader@26ba2a48

Error while starting bundle: file:/Users/nollymar/Documents/workspace-dotcms/tomcat8/webapps/ROOT/WEB-INF/felix/load/com.dotcms.override-0.2.jar: org.osgi.framework.BundleException: Activator start error in bundle com.dotcms.override [15].
org.osgi.framework.BundleException: Activator start error in bundle com.dotcms.override [15].
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2290)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2146)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316)
Caused by: java.lang.UnsupportedOperationException: class redefinition failed: attempted to add a method
	at sun.instrument.InstrumentationImpl.redefineClasses0(Native Method)
	at sun.instrument.InstrumentationImpl.redefineClasses(InstrumentationImpl.java:170)
	at net.bytebuddy.dynamic.loading.ClassReloadingStrategy$Strategy$1.apply(ClassReloadingStrategy.java:261)
	at net.bytebuddy.dynamic.loading.ClassReloadingStrategy.load(ClassReloadingStrategy.java:171)
	at net.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:79)
	at net.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:4376)
	at com.dotmarketing.osgi.GenericBundleActivator.overrideClass(GenericBundleActivator.java:383)
	at com.dotmarketing.osgi.GenericBundleActivator.overrideClass(GenericBundleActivator.java:348)
	at com.dotmarketing.osgi.GenericBundleActivator.overrideClasses(GenericBundleActivator.java:168)
	at com.dotmarketing.osgi.GenericBundleActivator.initializeServices(GenericBundleActivator.java:135)
	at com.dotmarketing.osgi.override.Activator.start(Activator.java:31)
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697)
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240)
	... 7 more
14:06:59.890  INFO  reindex.ReindexThread - --- ReindexThread Paused (every 3600000 millis)
```

**Comment by wezell** [comment]
@nollymar are you running with ByteBuddy installed as the agent?  Class overrides require that

**Comment by nollymar** [comment]
@wezell I am:

**Comment by nollymar** [comment]
@wezell I am. I set it as an argument of the startup script:

```
07-Jan-2021 14:03:00.767 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -javaagent:/Users/nollymar/Documents/workspace-dotcms/tomcat8/webapps/ROOT/WEB-INF/lib/byte-buddy-agent-1.6.12.jar
```

**Comment by wezell** [comment]
start of a test:

https://gist.github.com/wezell/630912956b0aa617ba9626eb1a1eb3a0

**Comment by nollymar** [comment]
Tests implemented here: https://github.com/dotCMS/core/pull/19817

**Comment by jcastro-dotcms** [comment]
The official code changes have been included in the dotCMS 5.3.8.4 LTS release only. The 5.2.8.4 LTS release is several commits behind, which made the code back-port not possible.
